packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/cluster_client.ts

```diff
@@ -78,28 +78,43 @@ export class ClusterClient implements ICustomClusterClient {
       kibanaVersion,
     });
     this.rootScopedClient = configureClient(config, {
+      scoped: true,
       logger,
       type,
       getExecutionContext,
-      scoped: true,
       agentFactoryProvider,
       kibanaVersion,
     });
   }
 
   asScoped(request: ScopeableRequest) {
-    const scopedHeaders = this.getScopedHeaders(request);
+    const createScopedClient = () => {
+      const scopedHeaders = this.getScopedHeaders(request);
 
-    const transportClass = createTransport({
-      getExecutionContext: this.getExecutionContext,
-      getUnauthorizedErrorHandler: this.createInternalErrorHandlerAccessor(request),
-    });
+      const transportClass = createTransport({
+        getExecutionContext: this.getExecutionContext,
+        getUnauthorizedErrorHandler: this.createInternalErrorHandlerAccessor(request),
+      });
+
+      return this.rootScopedClient.child({
+        headers: scopedHeaders,
+        Transport: transportClass,
+      });
+    };
+
+    const createSecondaryScopedClient = () => {
+      const secondaryAuthHeaders = this.getSecondaryAuthHeaders(request);
+
+      return this.asInternalUser.child({
+        headers: secondaryAuthHeaders,
+      });
```

Comment:
__question:__ I see we don't use a custom transport here, does it mean we won't support refresh & re-try for expired secondary authentication credentials (access token, when request is "real") like we do for `asCurrentUser`?

Reply:
Good point. 

Do you have any idea of what of error ES returns when the secondary auth is expired? Because I don't, but I suspect it's different than errors for the primary auth, right?

Also we won't be able to use the current logic of the custom transport, because it only handles primary authentication credentials, so it would mean move specific development for second auth. I'm not opposed doing so, but then it probably can be done in a later stage (in which case I will open an issue for it), would that be alright with you?

Reply:
> Do you have any idea of what of error ES returns when the secondary auth is expired? Because I don't, but I suspect it's different than errors for the primary auth, right?

I'd expect 401 for expired tokens even if the token is a part of the secondary auth, but it's a good question, we need to confirm. Let me see if I can quickly check it locally.

> Also we won't be able to use the current logic of the custom transport, because it only handles primary authentication credentials, so it would mean move specific development for second auth. I'm not opposed doing so, but then it probably can be done in a later stage (in which case I will open an issue for it), would that be alright with you?

Oh, I didn't realize we can't reuse the custom transport logic we already have. In any case, I think it's fine to treat it as an enhancement and handle it separately 👍

Reply:
> Oh, I didn't realize we can't reuse the custom transport logic we already have

I mean, looking at the code in createTransport

https://github.com/elastic/kibana/blob/35083504464a4950d5ce5c6464cab387dd98d0d8/packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/create_transport.ts#L59-L74

It should work "out of the box" on Core's side, but I think on the side of the security's handler we would need modifications, to be able to detect which token/header is expired to renew it and to then update the right header?

Reply:
>  but I think on the side of the security's handler we would need modifications, to be able to detect which token/header is expired to renew it and to then update the right header?

We might need to tweak our code slightly to remove the `es-secondary-authorization` header *before* re-authentication (see [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security/server/authentication/authentication_service.ts#L289-L296), unless Core does that for us). However, Security doesn't use headers from the request to extend the access token. Instead, we fetch the session for the request that Core provides us in the "unauthorized error handler" from ES and extract credentials to extend from there. So the only thing that matters to Security is what is located in the session.

The problem, I guess, is that Security will return a new access token within the `Authorization` HTTP header value. We need to decide whether Security should return it within the proper header, or if Core should take the value from `Authorization` and use it as `ES-Secondary-Authorization` instead. I'm leaning towards the latter since Core has a better chance of knowing that the request is performed via `asSecondaryAuthUser`, but we can discuss.

Reply:
> I'd expect 401 for expired tokens even if the token is a part of the secondary auth, but it's a good question, we need to confirm. Let me see if I can quickly check it locally.

Okay, checked:

1. Run ES with `yarn es snapshot --license trial --ssl -E xpack.security.authc.token.timeout=15s` to have short-lived tokens
2. Create an access token with
```http
POST https://localhost:9200/_security/oauth2/token
Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==
Content-Type: application/json
Accept: application/json

{
  "grant_type" : "client_credentials"
}
---
{
  "access_token": "wOSRBBj15MpZT8HbQdvoEkQFHAlcNC4J2gPqDZEAAAA=",
  "type": "Bearer",
  "expires_in": 15,
  ...
}
```
3. Request cluster state API while token is valid
```http
GET https://localhost:9200/_cluster/state
Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==
ES-Secondary-Authorization: Bearer wOSRBBj15MpZT8HbQdvoEkQFHAlcNC4J2gPqDZEAAAA=
---
200
---
{
  "cluster_name":"elasticsearch",
  ...
}
```
4. Request cluster state API after token has expired
```http
GET https://localhost:9200/_cluster/state
Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==
ES-Secondary-Authorization: Bearer wOSRBBj15MpZT8HbQdvoEkQFHAlcNC4J2gPqDZEAAAA=
---
401
---
{
  "error": {
    "root_cause": [
      {
        "type": "security_exception",
        "reason": "token expired",
        "header": {
          "WWW-Authenticate": "Bearer realm=\"security\", error=\"invalid_token\", error_description=\"The access token expired\""
        }
      }
    ],
    "type": "security_exception",
    "reason": "Failed to authenticate secondary user",
    "caused_by": {
      "type": "security_exception",
      "reason": "token expired",
      "header": {
        "WWW-Authenticate": "Bearer realm=\"security\", error=\"invalid_token\", error_description=\"The access token expired\""
      }
    }
  },
  "status": 401
}
```

The behavior seems to be exactly the same as for "primary" auth.

Reply:
"exactly the same" seems pretty bad for us though, given we'd ideally like a way to differentiate the two scenarios from within the reauth hook...

Reply:
> "exactly the same" seems pretty bad for us though, given we'd ideally like a way to differentiate the two scenarios from within the reauth hook...

Is it the custom transport that handles re-authentication/re-try on the Core side? I admit I don't know this code, so please correct me if I'm wrong, but if so, couldn't we create transports with different parameters for `asCurrentUser` and `asSecondaryAuthUser` to indicate which specific scenario we're handling?